labelindex/named_port_index.go

```diff
@@ -0,0 +1,690 @@
+		// Decref all the old members.  If they hit 0 references, then the member has been
+		// removed so we emit an event.
+		log.WithField("ipSetID", ipSetID).Debug("Removing endpoint from IP set")
+		ipSetData := idx.ipSetDataByID[ipSetID]
+		for _, oldMember := range contributions {
+			newRefCount := ipSetData.memberToRefCount[oldMember] - 1
+			if newRefCount == 0 {
+				// Member no longer in the IP set.  Emit event and clean up the old reference
+				// count.
+				idx.OnMemberRemoved(ipSetID, oldMember)
+				delete(ipSetData.memberToRefCount, oldMember)
+			} else {
+				ipSetData.memberToRefCount[oldMember] = newRefCount
+			}
+		}
+	}
+
+	// Record the new endpoint data.
+	delete(idx.endpointDataByID, id)
+	for _, parent := range oldEndpointData.parents {
+		parent.referenceCount--
+		idx.discardParentIfEmpty(parent.id)
+	}
+}
+
+func (idx *SelectorAndNamedPortIndex) UpdateParentLabels(parentID string, labels map[string]string) {
+	parentData := idx.getOrCreateParent(parentID)
+	if reflect.DeepEqual(parentData.labels, labels) {
+		log.WithField("parentID", parentID).Debug("Skipping no-op update to parent labels")
+		return
+	}
+	oldLabels := parentData.labels
+	idx.updateParent(
+		parentData,
+		// Function to apply the update.
+		func() {
+			parentData.labels = labels
+		},
+		// Function to back out the update.
+		func() {
+			parentData.labels = oldLabels
+		},
+	)
+}
+
+func (idx *SelectorAndNamedPortIndex) UpdateParentTags(parentID string, tags []string) {
+	parentData := idx.getOrCreateParent(parentID)
+	if reflect.DeepEqual(parentData.tags, tags) {
+		log.WithField("parentID", parentID).Debug("Skipping no-op update to parent labels")
+		return
+	}
+	oldTags := parentData.tags
+	idx.updateParent(
+		parentData,
+		// Function to apply the update.
+		func() {
+			parentData.tags = tags
+		},
+		// Function to back out the update.
+		func() {
+			parentData.tags = oldTags
+		},
+	)
+}
+
+func (idx *SelectorAndNamedPortIndex) updateParent(parentData *npParentData, applyUpdate, revertUpdate func()) {
+	for _, epData := range idx.endpointDataByID {
+		if !epData.HasParent(parentData) {
+			continue
+		}
+
+		// This endpoint matches this parent, calculate its old contribution.  (The revert function is a
+		// no-op on the first loop but keeping it here, rather than at the bottom of the loop makes it
+		// harder to accidentally skip it with a well-intentioned "continue".)
+		revertUpdate()
+		oldIPSetContributions := idx.RecalcCachedContributions(epData)
+
+		// Apply the update to the parent while we calculate this endpoint's new contribution.
+		applyUpdate()
+		for ipSetID, ipSetData := range idx.ipSetDataByID {
+			// Remove this IP set from the endpoint's cache of matching IP sets (if present).
+			// We'll restore it below if it still matches.
+			epData.RemoveMatchingIPSetID(ipSetID)
+
+			// Check if the selector matches.  Since we applied the update, this matches on the new
+			// values.
```

Comment:
The code here looks mostly the same as in UpdateEndpoint.  Can it be shared?

Reply:
Good spot, yes, I think it can